sdk/identity/identity/src/msal/msalClient.ts

```diff
@@ -0,0 +1,260 @@
+
+/**
+ * Interface for the MSAL (Microsoft Authentication Library) client.
+ * This client is used to interact with Microsoft's identity platform.
+ */
+export interface MsalClient {
+  /**
+   * Retrieves an access token by using a client secret.
+   *
+   * @param scopes - The scopes for which the access token is requested. These represent the resources that the application wants to access.
+   * @param clientSecret - The client secret of the application. This is a credential that the application can use to authenticate itself.
+   * @param options - Additional options that may be provided to the method.
+   * @returns An access token
+   */
+  getTokenByClientSecret(
+    scopes: string[],
+    clientSecret: string,
+    options?: GetTokenOptions,
+  ): Promise<AccessToken>;
+}
+
+/**
+ * Options for creating an instance of the MsalClient.
+ */
+export type CreateMsalClientOptions = Partial<Omit<MsalNodeOptions, "clientId" | "tenantId">>;
+
+/**
+ * Generates the configuration for MSAL (Microsoft Authentication Library).
+ *
+ * @param clientId - The client ID of the application.
+ * @param  tenantId - The tenant ID of the Azure Active Directory.
+ * @param  createMsalClientOptions - Optional. Additional options for creating the MSAL client.
+ * @returns  The MSAL configuration object.
+ */
+export function generateMsalConfiguration(
+  clientId: string,
+  tenantId: string,
+  createMsalClientOptions: CreateMsalClientOptions = {},
+): msal.Configuration {
+  const resolvedTenant = resolveTenantId(msalLogger, tenantId, clientId);
+
+  // TODO: move and reuse getIdentityClientAuthorityHost
+  const authority = getAuthority(
+    resolvedTenant,
+    createMsalClientOptions.authorityHost ?? process.env.AZURE_AUTHORITY_HOST,
+  );
+
+  const httpClient = new IdentityClient({
+    ...createMsalClientOptions.tokenCredentialOptions,
+    authorityHost: authority,
+    loggingOptions: createMsalClientOptions.loggingOptions,
+  });
+
+  const msalConfig: msal.Configuration = {
+    auth: {
+      clientId,
+      authority,
+      knownAuthorities: getKnownAuthorities(
+        resolvedTenant,
+        authority,
+        createMsalClientOptions.disableInstanceDiscovery,
+      ),
+    },
+    system: {
+      networkClient: httpClient,
+      loggerOptions: {
+        loggerCallback: defaultLoggerCallback(createMsalClientOptions.logger ?? msalLogger),
+        logLevel: getMSALLogLevel(getLogLevel()),
+        piiLoggingEnabled: createMsalClientOptions.loggingOptions?.enableUnsafeSupportLogging,
+      },
+    },
+  };
+  return msalConfig;
+}
+
+/**
+ * Represents the state necessary for the MSAL (Microsoft Authentication Library) client to operate.
+ * This includes the MSAL configuration, cached account information, Azure region, and a flag to disable automatic authentication.
+ *
+ * @internal
+ */
+interface MsalClientState {
+  /** The configuration for the MSAL client. */
+  msalConfig: msal.Configuration;
```

Comment:
Trying to minimize state, and only push things down if they apply to all credentials. If they apply to _some_ credentials they can be passed in as params to the various `getTokenBy...` calls